va/caa.go

```diff
@@ -238,10 +250,28 @@ func (va *ValidationAuthorityImpl) validateCAASet(caaSet *CAASet, wildcard bool)
 	//
 	// Our CAA identity must be found in the chosen checkSet.
 	for _, caa := range records {
-		if extractIssuerDomain(caa) == va.issuerDomain {
-			va.stats.Inc("CAA.Authorized", 1)
-			return true, true
+		caaIssuerDomain, caaParameters := extractIssuerDomainAndParameters(caa)
+		if caaIssuerDomain != va.issuerDomain {
+			continue
+		}
+
+		if features.Enabled(features.CAAValidationMethods) {
+			// Check the validation-methods CAA parameter as defined
+			// in section 4 of the draft CAA ACME RFC:
+			// https://tools.ietf.org/html/draft-ietf-acme-caa-03
+			caaMethods, ok := caaParameters["validation-methods"]
+			if ok {
+				if challengeType == nil {
```

Comment:
Small nit: seems like this check can be done further up in the initial feature check, i.e. `if features.Enabled(features.CAAValidationMethods) && challengeType != nil `. If we don't have the challengeType (because we are processing old messages or w/e) we can't determine if the method is correct anyway.

Reply:
IMO there are two reasons not to do this - firstly, it changes the behavior of the code. Currently, if the feature is enabled, the `validation-methods` parameter is present and the `challengeType` is `nil`, we'll fail the CAA validation check. If we combine check then if the feature is enabled and the `challengeType` is `nil`, we'll skip the check and will continue even if the `validation-methods` parameter is present. I suspect this is a bad thing, since a future bug whereby the proto field is not correctly populated, will result in silent skipping of the CAA validation-methods check (on the flip side, it does mean that the RA needs to be upgraded before or in sync with the VA). Secondly, I dislike mixing functional tests with feature tests - this makes it harder to later correctly remove the feature flag and in this case, the only cost is a single map lookup.

Reply:
We typically enforce population of proto fields in our gRPC wrappers (i.e. `grpc/va-wrappers.go`). Because we need to be roll out this change without taking all of the boulder components down we don't do that in the initial change so that that this component can still process messages from older RAs that don't know about the new proto field. Once the this change is deployed we'll then submit a new one updating the wrappers to be strict about the field being populated.

Since we'll pull most of these conditional checks out at that point anyway it's not a real issue, just a small style nit.